Installing uuagc (0.9.52.2) via cabal new-install fails

This may be a bug in Cabal instead of in uuagc, but you'll know that better than I do. :)

When installing uuagc using the new-style commands, i.e. `cabal new-install uuagc`, the dependency resolution phase fails due to a conflicting dependency on Cabal. Somehow, it determines that uuagc, directly or indirectly, depends on both Cabal >= 2.5 and <1.25, which clearly is not going to work.

Installing with the command `cabal new-install uuagc --constraint='uuagc==0.9.52.2' --verbose=3` (where the constraint is added to prevent cabal from trying older versions of uuagc), the following dependency resolution logs are generated: [uuagc-conflicts.txt](https://github.com/UU-ComputerScience/uuagc/files/3193859/uuagc-conflicts.txt)

From some cursory looking around, I have not been able to find the directive that requires Cabal to have a version <1.25, but that is probably my unfamiliarity with the dependency tree.

Thanks!

System info:
- macOS 10.13.6, though this also fails on Arch Linux
- cabal-install version 2.4.1.0
- ghc version 8.6.4

I made the repo (https://github.com/UU-ComputerScience/uuagc) work for GHC 8.6.5 and Cabal 3.0. 

Here are the things I did in case someone else needs it:

All this changes are on directory `uuagc/trunk` of uuagc reposiroty that I forked and cloned locally.

First, I saw some warnings on `cabal check`
```
> cabal check
Warning: These warnings may cause trouble when distributing the package:
Warning: From version 1.24 cabal supports specifying explicit dependencies for
Custom setup scripts. Consider using cabal-version >= 1.24 and adding a
'custom-setup' section with a 'setup-depends' field that specifies the
dependencies of the Setup.hs script itself. The 'setup-depends' field uses the
same syntax as 'build-depends', so a simple example would be 'setup-depends:
base, Cabal'.
➜  trunk git:(master) cabal check
➜  trunk git:(master) ✗ cabal check
No errors or warnings could be found in the package.
➜  trunk git:(master) ✗ cabal build
Resolving dependencies...
cabal: Could not resolve dependencies:
[__0] trying: uuagc-0.9.52.2 (user goal)
[__1] trying: base-4.12.0.0/installed-4.1... (dependency of uuagc)
[__2] next goal: uuagc-diagrams (user goal)
[__2] rejecting: uuagc-diagrams-0.1.1.0 (conflict:
base==4.12.0.0/installed-4.1..., uuagc-diagrams => base>=4.7 && <4.8)
[__2] rejecting: uuagc-diagrams-0.1.0.0 (constraint from user target requires
==0.1.1.0)
[__2] fail (backjumping, conflict set: base, uuagc-diagrams)
After searching the rest of the dependency tree exhaustively, these were the
goals I've had most trouble fulfilling: uuagc, base, uuagc-diagrams
```

So, I added a `custom-setup` as the warnings recommended, I added this into `uuagc.cabal` file:
```
custom-setup
  setup-depends: base, Cabal
```

Then, when I tried to build, it was less errors:
```
> cabal build
Resolving dependencies...
cabal: Could not resolve dependencies:
[__0] trying: uuagc-0.9.52.2 (user goal)
[__1] trying: base-4.12.0.0/installed-4.1... (dependency of uuagc)
[__2] next goal: uuagc-diagrams (user goal)
[__2] rejecting: uuagc-diagrams-0.1.1.0 (conflict:
base==4.12.0.0/installed-4.1..., uuagc-diagrams => base>=4.7 && <4.8)
[__2] rejecting: uuagc-diagrams-0.1.0.0 (constraint from user target requires
==0.1.1.0)
[__2] fail (backjumping, conflict set: base, uuagc-diagrams)
After searching the rest of the dependency tree exhaustively, these were the
goals I've had most trouble fulfilling: uuagc, base, uuagc-diagrams
```

It is trying to build `uuagc-diagrams` but the base library was conflicting. I saw also that that library is located in the same repo. So, I jumped into `diagrams` and tried to build it. I got a lot of errors. 

I reviewed if we have any dependency for uuagc-diagrams in uuagc.cabal. I found nothing. I do not know why this lib is not described in the cabal file of uuagc.

I also tried to build ´cabal-plugin´, this is `uuagc-cabal` library. And I got the same error, a dependency over `uuagc-diagrams`. Again, the cabal file does not describe the uuagc-diagrams dependency. Humm.

I had the idea of modifying `uuagc-diagrams`, and for that I would need a way to use my local modified package. So, I added a `cabal.project` in the `uuagc` directory with this content:
```
packages: ./
          cabal-plugin/
          diagrams/
```

Then I tried to make to compile `uuagc-diagrams`. I modified the cabal file `uuagc-diagrams.cabal` to consider a wider `base` versions, I changed from `base >=4.7 && <4.8` to `base >=4.7 && <5`. 

With these, it proceeded to build, but failed again over some errors on kinds and types:
```
UU/UUAGC/Diagrams.hs:94:20: error:
    • Expected kind ‘((* -> *) -> Constraint) -> k0’,
        but ‘Diagram b’ has kind ‘*’
    • In the type ‘Diagram b R2’
      In the type declaration for ‘AGDiagram’
   |
94 | type AGDiagram b = Diagram b R2
   |                    ^^^^^^^^^^^^
```

Weirdly, I moved to `cabal-plugin` and the build just passed without errors!, and then when I did the same for `uuagc`, it also passed without errors. I got surprised of this. 
Some how, we do not need a build success for `uuagc-diagrams` to build `uuagc`. But, I got what I wanted. UUAGC build passed successfully!

I hope this helps others who want to use uuagc, it is an amazing tool.

--- a/.gitignore
+++ b/.gitignore
@@ -3,3 +3,4 @@
 *.dyn_o
 *.dyn_hi
 uuagc/trunk/dist/*
+dist-newstyle/

--- a/uuagc/trunk/cabal-plugin/src/Distribution/Simple/UUAGC/UUAGC.hs
+++ b/uuagc/trunk/cabal-plugin/src/Distribution/Simple/UUAGC/UUAGC.hs
@@ -6,7 +6,7 @@ module Distribution.Simple.UUAGC.UUAGC(uuagcUserHook,
                                        uuagcFromString
                                       ) where
 
-import Distribution.Simple.BuildPaths (autogenModulesDir)
+-- import Distribution.Simple.BuildPaths (autogenComponentModulesDir)
 import Debug.Trace
 import Distribution.Simple
 import Distribution.Simple.PreProcess
@@ -104,13 +104,13 @@ uuagcLibUserHook :: ([String] -> FilePath -> IO (ExitCode, [FilePath])) -> UserH
 uuagcLibUserHook uuagc = hooks where
   hooks = simpleUserHooks { hookedPreProcessors = ("ag", ag):("lag",ag):knownSuffixHandlers
                           , buildHook = uuagcBuildHook uuagc
-                          , sDistHook = uuagcSDistHook uuagc
+--                          , sDistHook = uuagcSDistHook uuagc
                           }
   ag = uuagc' uuagc
 
 originalPreBuild  = preBuild simpleUserHooks
 originalBuildHook = buildHook simpleUserHooks
-originalSDistHook = sDistHook simpleUserHooks
+--originalSDistHook = sDistHook simpleUserHooks
 
 putErrorInfo :: Handle -> IO ()
 putErrorInfo h = hGetContents h >>= hPutStr stderr
@@ -188,21 +188,21 @@ getOptionsFromClass classes fOpt =
                                                    ++ show fClass
                                                    ++ " is not defined."
 
-uuagcSDistHook :: ([String] -> FilePath -> IO (ExitCode, [FilePath]))
-     -> PackageDescription
-     -> Maybe LocalBuildInfo
-     -> UserHooks
-     -> SDistFlags
-     -> IO ()
-uuagcSDistHook uuagc pd mbLbi uh df = do
-  {-
-  case mbLbi of
-    Nothing -> warn normal "sdist: the local buildinfo was not present. Skipping AG initialization. Dist may fail."
-    Just lbi -> let classesPath = buildDir lbi </> agClassesFile
-                in commonHook uuagc classesPath pd lbi (sDistVerbosity df)
-  originalSDistHook pd mbLbi uh df
-  -}
-  originalSDistHook pd mbLbi (uh { hookedPreProcessors = ("ag", nouuagc):("lag",nouuagc):knownSuffixHandlers }) df  -- bypass preprocessors
+-- uuagcSDistHook :: ([String] -> FilePath -> IO (ExitCode, [FilePath]))
+--      -> PackageDescription
+--      -> Maybe LocalBuildInfo
+--      -> UserHooks
+--      -> SDistFlags
+--      -> IO ()
+-- uuagcSDistHook uuagc pd mbLbi uh df = do
+--   {-
+--   case mbLbi of
+--     Nothing -> warn normal "sdist: the local buildinfo was not present. Skipping AG initialization. Dist may fail."
+--     Just lbi -> let classesPath = buildDir lbi </> agClassesFile
+--                 in commonHook uuagc classesPath pd lbi (sDistVerbosity df)
+--   originalSDistHook pd mbLbi uh df
+--   -}
+--   originalSDistHook pd mbLbi (uh { hookedPreProcessors = ("ag", nouuagc):("lag",nouuagc):knownSuffixHandlers }) df  -- bypass preprocessors
 
 uuagcBuildHook
   :: ([String] -> FilePath -> IO (ExitCode, [FilePath]))

--- a/uuagc/trunk/diagrams/UU/UUAGC/Diagrams.hs
+++ b/uuagc/trunk/diagrams/UU/UUAGC/Diagrams.hs
@@ -1,6 +1,7 @@
 {-# LANGUAGE UndecidableInstances #-}
 {-# LANGUAGE FlexibleContexts #-}
 {-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GADTs #-}
 
 -----------------------------------------------------------------------------
 -- |
@@ -32,6 +33,7 @@ module UU.UUAGC.Diagrams
 import Diagrams.Prelude
 import Graphics.SVGFonts (textSVG_, Spacing (..), TextOpts (..), lin2, Mode (..))
 import Data.List (isPrefixOf)
+import System.IO.Unsafe (unsafePerformIO)
 
 -- | Construct a diagram for a full production, given its inherited attributes,
 --   name, synthesized attributes and children
@@ -51,8 +53,8 @@ child i n s = Child $ node False i n s []
 --   the shape of the arrow and can be 'shaftL', 'shaftR', 'shaftT', 'shaftB'
 --   of 'shaftD', or a special trial constructed with the diagrams library.
 agrule :: AGBackend b =>
-          Trail R2 -> String -> String -> AGDiagram b -> AGDiagram b
-agrule sh s1 s2 = connectPerim' (with & headLength .~ (Normalized 0.025) & arrowShaft .~ sh) n1 n2 (tb t1) (tb t2) where
+          Trail V2 Double -> String -> String -> AGDiagram b -> AGDiagram b
+agrule sh s1 s2 = connectPerim' (with & headLength .~ (normalized 0.025) & arrowShaft .~ sh) n1 n2 (tb t1) (tb t2) where
   t1 = "lhs." `isPrefixOf` s1
   t2 = "lhs." `isPrefixOf` s2
   n1 | '.' `notElem` s1 = s1 -- terminal
@@ -65,14 +67,14 @@ agrule sh s1 s2 = connectPerim' (with & headLength .~ (Normalized 0.025) & arrow
 -- | Construct an induced dependency arrow between two attributes, similar to
 --   'agrule' but with an explicit trial.
 indrule :: AGBackend b => String -> String -> AGDiagram b -> AGDiagram b
-indrule s1 s2 = connectPerim' (with & headLength .~ (Normalized 0.025) & arrowShaft .~ shaftB & shaftStyle %~ dashed . opacity 0.5) n1 n2 tb tb where
+indrule s1 s2 = connectPerim' (with & headLength .~ (normalized 0.025) & arrowShaft .~ shaftB & shaftStyle %~ dashed . opacity 0.5) n1 n2 tb tb where
   t = "lhs." `isPrefixOf` s1
   n1 = if t then s1 ++ ".syn" else s1 ++ ".inh"
   n2 = if t then s2 ++ ".inh" else s2 ++ ".syn"
   tb = if t then 90 @@ deg else 270 @@ deg
   dashed  = dashingN [0.01,0.01] 0
 
-shaftL, shaftR, shaftT, shaftB, shaftD :: Trail R2
+shaftL, shaftR, shaftT, shaftB, shaftD :: Trail V2 Double
 
 -- | Line that first moves left and then right
 shaftL = fromSegments [bezier3 (r2 (0.5,0.3)) (r2 (0.5,-0.3)) (r2 (1,0))]
@@ -81,26 +83,26 @@ shaftL = fromSegments [bezier3 (r2 (0.5,0.3)) (r2 (0.5,-0.3)) (r2 (1,0))]
 shaftR = fromSegments [bezier3 (r2 (0.5,-0.3)) (r2 (0.5,0.3)) (r2 (1,0))]
 
 -- | Top half of a circle
-shaftT = arcCW (0 @@ turn) (3/5 @@ turn)
+shaftT = arc xDir (-3/5 @@ turn)
 
 -- | Bottom half of a circle
-shaftB = arc (0 @@ turn) (2/5 @@ turn)
+shaftB = arc xDir (2/5 @@ turn)
 
 -- | Straight line
 shaftD = straightShaft
 
 
 -- A bit ugly, but now user doesn't need to import diagrams package for just the types
-type AGDiagram b = Diagram b R2
-class (Renderable (Path R2) b, Backend b R2) => AGBackend b where
-instance (Renderable (Path R2) b, Backend b R2) => AGBackend b
+type AGDiagram b = QDiagram b V2 Double Any
+class (Renderable (Path V2 Double) b, Backend b (V b) Double) => AGBackend b where
+instance (Renderable (Path V2 Double) b, Backend b (V b) Double) => AGBackend b
 
 
 attr :: AGBackend b =>
         String -> Bool -> (String -> String) -> AGDiagram b
 attr s t f = stack t (unitSquare # named (f s) # lc black) (text' 0.7 s) where
-  stack True  a b = beside          unitY  a (b === strutY 0.2)
-  stack False a b = beside (negateV unitY) a (strutY 0.2 === b)
+  stack True  a b = beside   unitY  a (b === strutY 0.2)
+  stack False a b = beside (-unitY) a (strutY 0.2 === b)
 
 -- | Helper function for drawing a node
 node :: AGBackend b =>
@@ -116,22 +118,25 @@ node top inh s syn ch = res # applyAll lines where
   hcats s = hcat' (with & sep .~ s)
   els = inhs ++ [lhs] ++ syns
   toprow = beside unitX (
-             beside (negateV unitX) lhs
+             beside (-unitX) lhs
                (hcats 0.3 inhs ||| strutX 0.3))
              (strutX 0.3 ||| hcats 0.3 syns)
   inhs = map (\i -> attr i top (\n -> name ++ "." ++ n ++ ".inh")) inh
   syns = map (\s -> attr s top (\n -> name ++ "." ++ n ++ ".syn")) syn
   alines = zipWith line (map getName els) (map getName $ tail els) # lc grey
   name = if top then "lhs" else s
-  lhs = beside (negateV unitY) (
+  lhs = beside (-unitY) (
           beside unitY
           (circle 0.5 # named name # lc grey)
           (if top then (text' 0.9 s === strutY 0.1) else mempty))
         (strutY 0.1 === text' 0.9 name)
 
+{-# NOINLINE lin2' #-}
+lin2' = unsafePerformIO lin2
+
 text' :: AGBackend b =>
          Double -> String -> AGDiagram b
-text' d s = (textSVG_ (TextOpts s lin2 INSIDE_H KERN False d d)) # lw none # fc black # centerX
+text' d s = (textSVG_ (TextOpts lin2' INSIDE_H KERN False d d) s) # lw none # fc black # centerX
 
 line :: (IsName n1, IsName n2, AGBackend b) =>
         n1 -> n2 -> AGDiagram b -> AGDiagram b

--- a/uuagc/trunk/diagrams/uuagc-diagrams.cabal
+++ b/uuagc/trunk/diagrams/uuagc-diagrams.cabal
@@ -16,6 +16,6 @@ library
   exposed-modules:     UU.UUAGC.Diagrams
   -- other-modules:       
   -- other-extensions:    
-  build-depends:       base >=4.7 && <4.8, diagrams-lib >= 1.1, SVGFonts >= 1.4
+  build-depends:       base >=4.7 && <4.15, diagrams-lib >= 1.4 && <1.5, SVGFonts >= 1.7 && <1.8
   -- hs-source-dirs:      
   default-language:    Haskell2010

--- a/uuagc/trunk/uuagc.cabal
+++ b/uuagc/trunk/uuagc.cabal
@@ -33,6 +33,8 @@ flag with-loag
    default: False
    manual: True
 
+custom-setup
+  setup-depends: base, Cabal
 
 executable uuagc
    build-depends: uuagc-cabal >= 1.0.2.0

